rts/motoko-rts/src/persistence.rs

```diff
@@ -0,0 +1,223 @@
+//! Orthogonal persistence support
+//!
+//! Persistent metadata table, located at 6MB, in the static partition space.
+
+pub mod compatibility;
+
+use motoko_rts_macros::ic_mem_fn;
+
+use crate::{
+    barriers::write_with_barrier,
+    constants::{KB, MB},
+    gc::incremental::State,
+    memory::Memory,
+    persistence::compatibility::memory_compatible,
+    rts_trap_with,
+    types::{Value, TAG_BLOB},
+};
+
+use self::compatibility::TypeDescriptor;
+
+#[cfg(feature = "ic")]
+const FINGERPRINT: [char; 32] = [
+    'M', 'O', 'T', 'O', 'K', 'O', ' ', 'O', 'R', 'T', 'H', 'O', 'G', 'O', 'N', 'A', 'L', ' ', 'P',
+    'E', 'R', 'S', 'I', 'S', 'T', 'E', 'N', 'C', 'E', ' ', '3', '2',
+];
+#[cfg(feature = "ic")]
+const VERSION: usize = 1;
+/// The `Value` representation in the default-initialized Wasm memory.
+/// The GC ignores this value since it is a scalar representation.
+const DEFAULT_VALUE: Value = Value::from_scalar(0);
+
+/// The persistent metadata stored at the defined location `METADATA_ADDRESS` in memory.
+/// Use a long-term representation by relying on C layout.
+/// The `Value` references belong to the GC root set and require forwarding pointer resolution.
+#[repr(C)]
+struct PersistentMetadata {
+    /// Predefined character sequence in the memory to double check the orthogonal persistence mode.
+    fingerprint: [char; 32],
+    /// Version of the orthogonal persistence. To be increased on every persistent memory layout modification.
+    version: usize,
+    /// Reference to the stable sub-record of the actor, comprising all stable actor fields. Set before upgrade.
+    /// Constitutes a GC root and requires pointer forwarding.
+    stable_actor: Value,
+    /// Reference to the stable type descriptor, serving for heap compatibility checks on upgrades.
+    /// The blob in the descriptor serves as a GC root and requires pointer forwarding.
+    stable_type: TypeDescriptor,
+    /// The state of the incremental GC including the partitioned heap description.
+    /// The GC continues work after upgrades.
+    incremental_gc_state: State,
+    /// Upgrade performance statistics: Total number of instructions consumed by the last upgrade.
+    upgrade_instructions: u64,
+}
+
+/// Location of the persistent metadata. Prereserved and fixed forever.
+const METADATA_ADDRESS: usize = 4 * MB + 512 * KB;
+/// The reserved maximum size of the metadata, contains a reserve for future extension of the metadata.
+const METADATA_RESERVE: usize = 512 * KB;
+
+// TODO: Include partition table in reserved space.
+pub const HEAP_START: usize = METADATA_ADDRESS + METADATA_RESERVE;
+
+const _: () = assert!(core::mem::size_of::<PersistentMetadata>() <= METADATA_RESERVE);
```

Comment:
How much slack have you got?

Reply:
The metadata occupies 3840 bytes, so slack is around 508KB. Maybe too much?